core/modules/simpletest/simpletest.module

```diff
@@ -366,6 +366,14 @@ function simpletest_test_get_all() {
       $groups = array();
       foreach ($files as $file) {
         $classes = backdrop_parse_info_file($file->uri, TRUE);
+        // If the .tests.info file contained no classes, then it puts a slew of
+        // hard-to-diagnose errors into the watchdog log when the testing page
+        // is loaded. So we'll check and put a more informative error here.
+        $class_values = array_values($classes);
+        if (empty($classes) || !is_array($class_values[0])) {
+          watchdog('simpletest', 'File %uri contains no testing classes.', array('%uri' => $file->uri), WATCHDOG_WARNING, l(t('See the Testing Framework documentation.'), 'https://docs.backdropcms.org/documentation/testing-framework'));
+          continue;
+        }
```

Comment:
Nitpicking, but its probably more likely that youll bail out because of an empty info file so we should run that first then check if theres an empty class, so maybe
```
        $classes = backdrop_parse_info_file($file->uri, TRUE);
        // Ensure test info file is not empty and that each entry is properly
        // formatted.
        if (empty($classes)) {
          watchdog('simpletest', 'File %uri contains no testing classes.', array('%uri' => $file->uri), WATCHDOG_WARNING, l(t('See the Testing Framework documentation.'), 'https://docs.backdropcms.org/documentation/testing-framework'));
          continue;
        }
        foreach ($classes as $class_name => $class) {
          $class_values = array_values($classes);
          if (!is_array($class)) {
            watchdog('simpletest', 'File %uri contains  a malformed test info entry for %class_name.', array('%uri' => $file->uri, '%class_name' => $class_name), WATCHDOG_WARNING, l(t('See the Testing Framework documentation.'), 'https://docs.backdropcms.org/documentation/testing-framework'));
            continue; // or maybe continue 2, depending on whats happening later in the code, I havent checked
          }
        }
```

Reply:
Actually, in the situations I'd encountered, it was the case that the file contains no testing classes, but that would go through the second branch (wrongly) in this example code. (This is the case with Search API module, for example.) 

What happens (happened) is that Coder Upgrade creates a completely empty file, but when the person creates a release, the packager adds the timestamp, etc., so the file is no longer empty, and that leads to the second situation. So it's not that the file contains a malformed class; it's that it truly contains no classes at all, but `backdrop_parse_info()` is what creates the appearance of a malformed class.

The current (simpler) code covers both situations, and gives the right message: "go look at your .info file", so I suggest we leave this as it currently is.

Reply:
The situation *should* be that contrib should be responsible to ensure that `test.info` files are properly formatted. But in this case we find that if contrib messes up, in any way, core doesnt properly warn us about which contrib is messing up, so we should fix core this time. So I feel the solution should seek to discover the possible ways contrib could mess up and mitigate that. The solution youve got is specifically for empty files. But we get errors in these situations:
- files with statements (xxx = yyy) only
- file with statements before test cases
- incomplete sections (missing either or all of name, description, group, or file)

In any of those situations the user would have no idea where this error is coming from, which is what I think we should try and solve, not just one situation.

FWIW my "solution" above was untested, and wouldnt actually sove any of what I just described.

Reply:
Those are all good cases to trap. I'll put together a suite of files malformed in those (and any other ways I can think of) and revise the PR to handle them. Copying the relevant bits of your comment into the issue for visibility.